## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.

Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [ ]:
# reading the files and storing the files
try:
    df = pd.read_csv("/datasets/movies_and_shows.csv")
except FileNotFoundError:
    df = pd.read_csv('https://practicum-content.s3.us-west-1.amazonaws.com/datasets/movies_and_shows.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


Print the first 10 table rows:

In [3]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
  
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [4]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'.


### Conclusions <a id='data_review_conclusions'></a>

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
# the list of column names in the df table
df.columns.tolist()

['   name',
 'Character',
 'r0le',
 'TITLE',
 '  Type',
 'release Year',
 'genres',
 'imdb sc0re',
 'imdb v0tes']

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
# renaming columns
df.rename(columns= {'   name': 'name',
 'Character': 'character',
 'r0le': 'role',
 'TITLE': 'title',
 '  Type': 'type',
 'release Year': 'release_year',
 'imdb sc0re': 'imdb_score',
 'imdb v0tes': 'imdb_votes'},inplace=True)

In [7]:
df

name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
85574    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
85575  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
85576        In??s Prieto                    Fanny     ACTOR      Lokillo   
85577        Isabel Gaona                   Cacica     ACTOR      Lokillo   
85578      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes  
0          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
1          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
2          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
3          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
4          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
...          ...           ...                 ...         ...         ...  
85574  the movie          2021          ['comedy']         3.8        68.0  
85575  the movie          2021          ['comedy']         3.8        68.0  
85576  the movie          2021          ['comedy']         3.8        68.0  
85577      MOVIE          2021          ['comedy']         3.8        68.0  
85578  the movie          2021          ['comedy']         3.8        68.0  

[85579 rows x 9 columns]

Check the result. Print the names of the columns once more:

In [8]:
# checking result: the list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [9]:
# calculating missing values
df.isnull().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [10]:
# dropping rows where columns with title, scores and votes have missing values
df.dropna(subset=['imdb_score', 'title', 'imdb_votes'], inplace=True)

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [11]:
# counting missing values
df.isnull().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [12]:
# counting duplicate row
df.duplicated().sum()

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [13]:
# Produce table with duplicates (with original rows included) and review last 5 rows


duplicates = df[df.duplicated(keep=False)]

print("Table with duplicates:")
display(duplicates)

print("\nLast 5 Rows of duplicates:")
display(duplicates.tail())

Table with duplicates:


name                  character      role  \
7560         Philip Greene  Baseball Fan (uncredited)     ACTOR   
7561         Philip Greene  Baseball Fan (uncredited)     ACTOR   
14502             Dan Levy                   Reporter     ACTOR   
14512             Dan Levy                   Reporter     ACTOR   
18951     Nicolas Le Nev??                    unknown  DIRECTOR   
...                    ...                        ...       ...   
85569       Jessica Cediel            Liliana Navarro     ACTOR   
85570  Javier Gardeaz?­bal   Agust??n "Peluca" Ort??z     ACTOR   
85571        Carla Giraldo             Valery Reinoso     ACTOR   
85572  Ana Mar??a S?­nchez                    Lourdes     ACTOR   
85577         Isabel Gaona                     Cacica     ACTOR   

                                 title   type  release_year  \
7560                   How Do You Know  MOVIE          2010   
7561                   How Do You Know  MOVIE          2010   
14502  A Very Harold & Kumar Christmas  MOVIE          2011   
14512  A Very Harold & Kumar Christmas  MOVIE          2011   
18951                       Sammy & Co   SHOW          2014   
...                                ...    ...           ...   
85569                          Lokillo  MOVIE          2021   
85570                          Lokillo  MOVIE          2021   
85571                          Lokillo  MOVIE          2021   
85572                          Lokillo  MOVIE          2021   
85577                          Lokillo  MOVIE          2021   

                                 genres  imdb_score  imdb_votes  
7560     ['comedy', 'drama', 'romance']         5.4     50383.0  
7561     ['comedy', 'drama', 'romance']         5.4     50383.0  
14502  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
14512  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
18951         ['animation', 'european']         5.7        31.0  
...                                 ...         ...         ...  
85569                        ['comedy']         3.8        68.0  
85570                        ['comedy']         3.8        68.0  
85571                        ['comedy']         3.8        68.0  
85572                        ['comedy']         3.8        68.0  
85577                        ['comedy']         3.8        68.0  

[13988 rows x 9 columns]


Last 5 Rows of duplicates:


name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imdb_score  imdb_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80853 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          80853 non-null  object 
 1   character     80853 non-null  object 
 2   role          80853 non-null  object 
 3   title         80853 non-null  object 
 4   type          80853 non-null  object 
 5   release_year  80853 non-null  int64  
 6   genres        80853 non-null  object 
 7   imdb_score    80853 non-null  float64
 8   imdb_votes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 6.2+ MB


In [15]:
# removing duplicate rows

df = df.drop_duplicates() # only display the correct answer, but the orignal df is still doesn't change

Check for duplicate rows once more to make sure you have removed all of them:

In [16]:
# checking for duplicates only show explicit duplicates

df.duplicated().sum()

0

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [17]:
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

In [18]:
# viewing unique type names

unique_types = df['type'].unique()



In [19]:
unique_types

array(['MOVIE', 'the movie', 'tv show', 'shows', 'movies', 'tv shows',
       'tv series', 'tv', 'SHOW'], dtype=object)

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters:
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [20]:
# function for replacing implicit duplicates
def replace_wrong_show(df, wrong_list, correct_value):
    df.loc[:, 'type'] = df['type'].replace(wrong_list, correct_value)
    return df


Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [21]:
# removing implicit duplicates
wrong_list = ['shows','SHOW','tv show','tv shows','tv series','tv']
correct_value = 'SHOW'

df = replace_wrong_show(df, wrong_list, correct_value)


/opt/conda/envs/python3/lib/python3.9/site-packages/pandas/core/indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [22]:
# viewing unique genre names
unique_genres = df['type'].unique()
print("Unique Genre Names:")
print(unique_genres)

Unique Genre Names:
['MOVIE' 'the movie' 'SHOW' 'movies']


In [23]:
df['type'].unique()

array(['MOVIE', 'the movie', 'SHOW', 'movies'], dtype=object)

In [24]:
df.duplicated().sum()

43

In [25]:
df = df.drop_duplicates()
df.duplicated().sum()

0

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed.

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [26]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
df_slice = df[df['release_year'] >= 1999]
# give the slice of dataframe new name
new_df = df_slice.copy()

In [27]:
# repeat conditional indexing so df has only shows (movies are removed as result)
df_slice = df_slice[df_slice['type'] == 'SHOW'] # how to make conditional slice

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [28]:
# rounding column with scores
df_slice = df_slice.round(0)

#checking the outcome with tail()
df_slice.tail()

name character      role      title  type  release_year  \
77655     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  SHOW          2021   
77656    Rudklao Amratisha   unknown     ACTOR  Let's Eat  SHOW          2021   
77657      Jaturong Mokjok   unknown     ACTOR  Let's Eat  SHOW          2021   
77658     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  SHOW          2021   
77659  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  SHOW          2021   

                    genres  imdb_score  imdb_votes  
77655  ['drama', 'comedy']         8.0         5.0  
77656  ['drama', 'comedy']         8.0         5.0  
77657  ['drama', 'comedy']         8.0         5.0  
77658  ['drama', 'comedy']         8.0         5.0  
77659  ['drama', 'comedy']         8.0         5.0

It is now time to identify outliers based on the number of votes.

In [29]:
# Use groupby() for scores and count all unique values in each group, print the result

result = df_slice.groupby('imdb_score')['name'].count()

print(result)

imdb_score
2.0       24
3.0       27
4.0      179
5.0      589
6.0     2477
7.0     4706
8.0     4820
9.0      557
10.0       8
Name: name, dtype: int64


Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [30]:
# filter dataframe using two conditions (scores to be in the range 4-9)
filtered_df = df_slice.loc[(df_slice['imdb_score'] >= 4) & (df_slice['imdb_score'] <= 9)]
# group scores and corresponding average number of votes, reset index and print the result
grouped_df = filtered_df.groupby('imdb_score', as_index=False)['imdb_votes'].mean()

grouped_df.reset_index(drop=True, inplace=True)

grouped_df

imdb_score     imdb_votes
0         4.0    5296.553073
1         5.0    3159.726655
2         6.0    3503.104966
3         7.0    8727.068211
4         8.0   30431.420539
5         9.0  126904.109515

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [31]:
# round column with averages
grouped_df = grouped_df.round() #the code should be precise
# rename columns
grouped_df.rename(columns={'imdb_score':'Score' ,'imdb_votes': 'Average_Votes' }, inplace=True)
# print dataframe in descending order
grouped_df = grouped_df.sort_values(by='Average_Votes', ascending=False)

grouped_df

Score  Average_Votes
5    9.0       126904.0
4    8.0        30431.0
3    7.0         8727.0
0    4.0         5297.0
2    6.0         3503.0
1    5.0         3160.0

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.